pkg/client/cache/reflector.go

```diff
@@ -160,16 +164,50 @@ var (
 // required, and a cleanup function.
 func (r *Reflector) resyncChan() (<-chan time.Time, func() bool) {
 	if r.resyncPeriod == 0 {
+		r.nextResync = time.Time{}
 		return neverExitWatch, func() bool { return false }
 	}
 	// The cleanup function is required: imagine the scenario where watches
 	// always fail so we end up listing frequently. Then, if we don't
 	// manually stop the timer, we could end up with many timers active
 	// concurrently.
+	r.nextResync = time.Now().Add(r.resyncPeriod)
 	t := time.NewTimer(r.resyncPeriod)
 	return t.C, t.Stop
 }
 
+// We want to avoid situations when periodic resyncing is breaking the TPC
+// connection.
+// If response`s body is not read to completion before calling body.Close(),
+// that TPC connection will not be reused in the future - see #15664 issue
+// for more details.
+// Thus, we set timeout for watch requests to be smaller than the remaining
+// time until next periodic resync and force resyncing ourself to avoid
+// breaking TCP connection.
+func (r *Reflector) timeoutForWatch() *int64 {
+	randTimeout := time.Duration(float64(5*time.Minute) * (rand.Float64() + 1.0))
```

Comment:
Yes - this is generally to spread the load.
So to be more precise - currently if timeout is not set explicitly, we are setting timeout to spread the load here:
https://github.com/kubernetes/kubernetes/blob/master/pkg/apiserver/resthandler.go#L281

Since this will be setting explicit timeout, that logic will not be invoked, but I want to leave that logic (also, because we want to avoid very long requests (e.g. hours) when resync period is very long).

So what do you expect to be a parameter? This 5min? Is it wort doing it basing on the fact that (I think) noone ever change that 5 minutes in apiserver?